lib/puppet/functions/each.rb

```diff
@@ -1,38 +1,76 @@
-#  Applies a parameterized block to each element in a sequence of selected entries from the first
-#  argument and returns the first argument.
+# Runs a [lambda](http://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html)
+# repeatedly using each value in a data structure, then returning the values unchanged.
 #
-# This function takes two mandatory arguments: the first should be an Array or a Hash or something that is
-# of enumerable type (integer, Integer range, or String), and the second
-# a parameterized block as produced by the puppet syntax:
+# This function takes two mandatory arguments, in this order:
 #
-#       $a.each |$x| { ... }
-#       each($a) |$x| { ... }
+# 1. An array or hash the function will iterate over.
+# 2. A lambda, which the function calls for each element in the first argument. It can
+# request one or two parameters.
 #
-# When the first argument is an Array (or of enumerable type other than Hash), the parameterized block
-# should define one or two block parameters.
-# For each application of the block, the next element from the array is selected, and it is passed to
-# the block if the block has one parameter. If the block has two parameters, the first is the elements
-# index, and the second the value. The index starts from 0.
+# @example Using the `each` function
 #
-#       $a.each |$index, $value| { ... }
-#       each($a) |$index, $value| { ... }
+# `$each_data = $data.each |$parameter| { <PUPPET CODE BLOCK> }`
```

Comment:
Okay, here's an example of how the iteration functions are all slightly different. With `each`, you'll _never_ want to capture the return value, because you already have it (in the same `$data` variable you're iterating over). It works mostly the same way `map` does, except that you're generally calling it solely for the side effects of the code in the lambda. 

Short version is that you're almost always writing resource declarations in an `each` lambda, and the point of calling `each` is to add those resources to the catalog.